liquidhaskell-boot/src/Language/Haskell/Liquid/Constraint/RewriteCase.hs

```diff
@@ -0,0 +1,127 @@
+module Language.Haskell.Liquid.Constraint.RewriteCase 
+    (getCaseRewrites) 
+    where
+
+import           Language.Fixpoint.Types
+import qualified Language.Fixpoint.Misc as M
+import           Language.Haskell.Liquid.Constraint.Types
+import           Language.Haskell.Liquid.Types.Types
+import           Language.Haskell.Liquid.Types.RType
+
+import           Data.Maybe
+import           Data.Tuple
+
+import qualified Data.HashMap.Strict as M
+import qualified Data.HashSet        as S
+
+getCaseRewrites :: CGEnv -> SpecType -> LocalRewrites
+getCaseRewrites γ spec =
+    let Reft (_, refinement) = ur_reft $ rt_reft spec
+        -- Names of constustors both type constructors and data constructors
+        ctors   = toSet $ seBinds  $ constEnv γ
+        -- All the global names (top level functions, types, etc...)
+        globals = toSet $ reGlobal $ renv     γ
+    in unloop
+       $ concatMap (uncurry $ unify ctors globals)
+       $ groupUnifiableEqualities 
+       $ getEqualities refinement
+    where toSet = S.fromList . M.keys
+
+-- We don't need the a total unification as constructors are inejctive
```

Comment:
What would be total unification?

Reply:
That unification success only if we can unify everything